server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -977,6 +1045,175 @@ private void insertPendingSegmentIntoMetastore(
+                                 .add(segment.getInterval());
+      committedIntervalToSegments.computeIfAbsent(segment.getInterval(), i -> new HashSet<>())
+                                 .add(segment);
+    }
+
+    final Set<DataSegment> upgradedSegments = new HashSet<>();
+    for (Map.Entry<String, Set<Interval>> entry : committedVersionToIntervals.entrySet()) {
+      final String upgradeVersion = entry.getKey();
+      Map<Interval, Set<DataSegment>> segmentsToUpgrade = getSegmentsWithVersionLowerThan(
+          upgradeVersion,
+          entry.getValue(),
+          appendVersionToSegments
+      );
+      for (Map.Entry<Interval, Set<DataSegment>> upgradeEntry : segmentsToUpgrade.entrySet()) {
+        Set<DataSegment> segmentsUpgradedToVersion = upgradeSegmentsToVersion(
+            handle,
+            upgradeVersion,
+            upgradeEntry.getKey(),
+            upgradeEntry.getValue(),
+            committedIntervalToSegments
+        );
+        log.info("Upgraded [%d] segments to version[%s].", segmentsUpgradedToVersion.size(), upgradeVersion);
+        upgradedSegments.addAll(segmentsUpgradedToVersion);
+      }
+    }
+
+    return upgradedSegments;
+  }
+
+  /**
+   * Creates a Map from eligible interval to Set of segments that are fully
+   * contained in that interval and have a version strictly lower than {@code #cutoffVersion}.
+   */
+  private Map<Interval, Set<DataSegment>> getSegmentsWithVersionLowerThan(
+      String cutoffVersion,
+      Set<Interval> eligibleIntervals,
+      TreeMap<String, Set<DataSegment>> versionToSegments
+  )
+  {
+    final Set<DataSegment> eligibleSegments
+        = versionToSegments.headMap(cutoffVersion).values().stream()
+                           .flatMap(Collection::stream)
+                           .collect(Collectors.toSet());
+
+    final Map<Interval, Set<DataSegment>> eligibleIntervalToSegments = new HashMap<>();
+
+    for (DataSegment segment : eligibleSegments) {
+      final Interval segmentInterval = segment.getInterval();
+      for (Interval eligibleInterval : eligibleIntervals) {
+        if (eligibleInterval.contains(segmentInterval)) {
+          eligibleIntervalToSegments.computeIfAbsent(eligibleInterval, itvl -> new HashSet<>())
+                                    .add(segment);
+          break;
+        } else if (eligibleInterval.overlaps(segmentInterval)) {
+          // Committed interval overlaps only partially
+          throw new ISE(
+              "Committed interval[%s] conflicts with interval[%s] of append segment[%s].",
+              eligibleInterval, segmentInterval, segment.getId()
+          );
+        }
+      }
+    }
+
+    return eligibleIntervalToSegments;
+  }
+
+  /**
+   * Computes new Segment IDs for the {@code segmentsToUpgrade} being upgraded
+   * to the given {@code upgradeVersion}.
+   */
+  private Set<DataSegment> upgradeSegmentsToVersion(
```

Comment:
the name is a bit confusing because it doesn't actually upgrade anything yet. Just creating extra versions.

Reply:
How about `createUpgradedVersionOfSegments`?

Reply:
that works too.